lib/Conversion/TorchToLinalg/Pooling.cpp

```diff
@@ -626,230 +626,68 @@ This is problematic for linalg ops for a few reasons:
 h! Although it is a bit like using a hammer to paint, our workaround is to use
 tensor.extract to access the elements of the input tensor inside our linalg
 generic op's payload.
-
-Current TODO's:
-    1. gather most of the boilerplate out of this op and make it into an
-adaptive pooling helper function.
-    2. figure out what to do with the conflicting decompositions in
-DecomposeComplexOps.cpp
-    3. Implement more efficient passes for when the kernel-size, input spatial
-dims, and output spatial dims are constant.
 */
 
+// TODO: infer template Dim from OpTy
 namespace {
-class ConvertAtenAdaptiveAvgPool1dOp
-    : public OpConversionPattern<AtenAdaptiveAvgPool1dOp> {
+template <typename OpTy, int Dim>
+class ConvertAtenAdaptivePoolOp : public OpConversionPattern<OpTy> {
 public:
-  using OpConversionPattern::OpConversionPattern;
+  using OpConversionPattern<OpTy>::OpConversionPattern;
   LogicalResult
-  matchAndRewrite(AtenAdaptiveAvgPool1dOp op, OpAdaptor adaptor,
+  matchAndRewrite(OpTy op, typename OpTy::Adaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
 
-    Location loc = op->getLoc();
-    const TypeConverter *typeConverter = getTypeConverter();
+    // check if we are doing a max pool or avg pool.
+    constexpr bool isMaxPool =
+        (llvm::is_one_of<OpTy, AtenAdaptiveMaxPool1dOp, AtenAdaptiveMaxPool2dOp,
+                         AtenAdaptiveMaxPool3dOp>::value);
+    static_assert(
+        isMaxPool ||
+            llvm::is_one_of<OpTy, AtenAdaptiveAvgPool1dOp,
+                            AtenAdaptiveAvgPool2dOp, AtenAdaptiveAvgPool3dOp,
+                            Aten_AdaptiveAvgPool3dOp>::value,
```

Comment:
> AdaptiveAvgPool3dOp appears twice here

In fact, Aten_AdaptiveAvgPool3dOp and AtenAdaptiveAvgPool3dOp are different. For some reason, dynamo was tracing the former rather than the latter, which was the reason for the cryptic torchdynamo failure mentioned in my previous comment. Frankly, it might be a good idea to add support for the other underscored ops just in case, i.e., Dim = 1 and 2.

> Rather do
> 
> constexpr bool isAvgPool = (llvm:: ... 
> static_assert(isMaxPool || isAvgPool, ... );

Yeah, that looks neater.

Reply:
> In fact, Aten_AdaptiveAvgPool3dOp and AtenAdaptiveAvgPool3dOp are different
Ouch, that's sure to confuse future me. I'll try to remember that.